src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs

```diff
@@ -0,0 +1,204 @@
+                {
+                    buildId = await this.GetBuildIdAsync(context, buildPipelineDefinition, buildVersionToDownload, project, tagsInput);
+                }
+                else if (buildVersionToDownload == buildVersionToDownloadSpecific)
+                {
+                    buildId = Int32.Parse(userSpecifiedBuildId);
+                }
+                else if (buildVersionToDownload == buildVersionToDownloadLatestFromBranch)
+                {
+                    buildId = await this.GetBuildIdAsync(context, buildPipelineDefinition, buildVersionToDownload, project, tagsInput, branchName);
+                }
+                else
+                {
+                    throw new InvalidOperationException("Unreachable code!");
+                }
+                string fullPath = this.CreateDirectoryIfDoesntExist(downloadPath);
+
+                context.Output(StringUtil.Loc("DownloadArtifactTo", downloadPath));
+                var server = new PipelineArtifactServer();
+                await server.DownloadAsyncWithProjectNameMiniMatch(context, project, buildId, artifactName, downloadPath, minimatchPatterns, token);
+                context.Output(StringUtil.Loc("DownloadArtifactFinished"));
+            }
+            else
+            {
+                throw new InvalidOperationException("Unreachable code!");
+            }
+        }
+
+        private string CreateDirectoryIfDoesntExist(string downloadPath)
+        {
+            string fullPath = Path.GetFullPath(downloadPath);
+            bool dirExists = Directory.Exists(fullPath);
+            if (!dirExists)
+            {
+                Directory.CreateDirectory(fullPath);
+            }
+            return fullPath;
+        }
+
+        private async Task<int> GetBuildIdAsync(AgentTaskPluginExecutionContext context, string buildPipelineDefinition, string buildVersionToDownload, string project, string[] tagFilters, string branchName=null)
+        {
+            var definitions = new List<int>() { Int32.Parse(buildPipelineDefinition) };
+            VssConnection connection = context.VssConnection;
+            BuildHttpClient buildHttpClient = connection.GetClient<BuildHttpClient>();
+            List<Build> list;
+            if (buildVersionToDownload == "latest")
+            {
+                list = await buildHttpClient.GetBuildsAsync(project, definitions, tagFilters: tagFilters, queryOrder: BuildQueryOrder.FinishTimeDescending);
+            }
+            else if (buildVersionToDownload == "latestFromBranch")
+            {
+                list = await buildHttpClient.GetBuildsAsync(project, definitions, branchName: branchName, tagFilters: tagFilters, queryOrder: BuildQueryOrder.FinishTimeDescending);
+            }
+            else
+            {
+                throw new InvalidOperationException("Unreachable code!");
+            }
+
+            if (list.Any())
+            {
+                return list.First().Id;
```

Comment:
do we ever expect `list` contains more than one elements? should we assert it?

Reply:
We expect it as we are querying for multiple builds given a set filters.